utils/net_utils/src/download.rs

```diff
@@ -0,0 +1,103 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use async_std::{
+    fs,
+    io::{self, copy, Read as AsyncRead},
+};
+use futures::task::{Context, Poll};
+use indicatif::{ProgressBar, ProgressStyle};
+use isahc::HttpClient;
+use log::info;
+use std::{marker::Unpin, path::Path, pin::Pin};
+use thiserror::Error;
+use url::Url;
+
+/// Contains progress bar and reader.
+struct DownloadProgress<S>
+where
+    S: Unpin,
+{
+    inner: S,
+    progress_bar: ProgressBar,
+}
+
+#[derive(Debug, Error)]
+enum DownloadError {
+    #[error("Cannot read a file header")]
+    HeaderError,
+    #[error("Filename encoding error")]
+    EncodingError,
+}
+
+impl<S: AsyncRead + Unpin> AsyncRead for DownloadProgress<S> {
+    fn poll_read(
```

Comment:
implementing synchronous `Read` would be another option, I initially implemented `AsyncRead` when attempting to use `reqwest` as `reqwest` only implements `AsyncRead` and stick to it since.

Reply:
Reason for not reusing logic from paramfetch which does the same as this?

Reply:
oh, didn't know it existed. However, `FetchProgress` has `inner` defined as `ProgressBar<prb::Pipe>`, and it's a multibar, in my case I wanted to use single bar and have `inner` as `ProgressBar<Stdout>`. 

Attempted to make `inner` be generic enough to cover both types, ie `ProgressBar<W: Write>`, but didn't really succeed so far.

Reply:
actually, nevermind, `ProgressBar<W>` works well enough!